### PR TITLE
bind capabilities if actor has it in its claims

### LINKED ISSUE #239: `wascc/hello-world-assemblyscript`: actor is not authorized to use capability wascc:logging
This may just be me and/or a limitation of my attempt to run krustlet w/ MicroK8s.

```bash
kubectl apply --filename=https://raw.githubusercontent.com/deislabs/krustlet/master/demos/wascc/hello-world-assemblyscript/k8s.yaml
pod/hello-world-wascc-assemblyscript configured

kubectl describe pod/hello-world-wascc-assemblyscript
Name:         hello-world-wascc-assemblyscript
Namespace:    default
Priority:     0
Node:         krustlet/
Labels:       <none>
Annotations:  Status:  Failed
Message:      Failed to run pod: Error configuring capabilities for module: WebAssembly module authorization failure: Unauthorized binding: actor MA6DECXXYRIRPWYJAIKQT7ZBPUCJ7HRBLK3ZSDTGE7WWIJ6VU52PZOCI is not authorized to use capability wascc:logging.
IP:           
IPs:          <none>
Containers:
  hello-world-wascc-assemblyscript:
    Container ID:   
    Image:          webassembly.azurecr.io/hello-world-wascc-assemblyscript:v0.1.0
    Image ID:       
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Terminated
      Message:      Error while starting container: Error configuring capabilities for module: WebAssembly module authorization failure: Unauthorized binding: actor MA6DECXXYRIRPWYJAIKQT7ZBPUCJ7HRBLK3ZSDTGE7WWIJ6VU52PZOCI is not authorized to use capability wascc:logging.
      Exit Code:    1
      Started:      Mon, 01 Jan 0001 00:00:00 +0000
      Finished:     Fri, 08 May 2020 10:36:50 -0700
    Ready:          False
    Restart Count:  0
    Environment:
      PORT:  8080
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-x75t7 (ro)
Conditions:
  Type           Status
  PodScheduled   True 
Volumes:
  default-token-x75t7:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-x75t7
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  beta.kubernetes.io/arch=wasm32-wascc
Tolerations:     krustlet/arch=wasm32-wascc:NoExecute
                 node.kubernetes.io/network-unavailable:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age        From               Message
  ----    ------     ----       ----               -------
  Normal  Scheduled  <unknown>  default-scheduler  Successfully assigned default/hello-world-wascc-assemblyscript to krustlet
```

The Krustlet, run as:

```bash
RUST_LOG=info \
KUBECONFIG=${HOME}/.kube/config \
./target/debug/krustlet-wascc \
--node-ip=127.0.0.1 \
--node-name=krustlet \
--pfx-password=password \
--pfx-path=./krustlet.pfx \
--port=3000
```
Outputs:
```bash
[2020-05-08T17:36:50Z INFO  wascc_provider] configuring capability wascc:http_server
[2020-05-08T17:36:50Z INFO  wascc_host] Attempting to bind actor MA6DECXXYRIRPWYJAIKQT7ZBPUCJ7HRBLK3ZSDTGE7WWIJ6VU52PZOCI to default,wascc:http_server
[2020-05-08T17:36:50Z INFO  wascc_httpsrv] Received HTTP Server configuration for MA6DECXXYRIRPWYJAIKQT7ZBPUCJ7HRBLK3ZSDTGE7WWIJ6VU52PZOCI
[2020-05-08T17:36:50Z INFO  wascc_provider] configuring capability wascc:logging
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 98, kind: AddrInUse, message: "Address already in use" }', /home/dazwilkin/.cargo/registry/src/github.com-1ecc6299db9ec823/wascc-httpsrv-0.6.0/src/lib.rs:92:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It appears (!) to successfully use the `wascc:http_server` capability but fail on `wascc:logging`.

### PR DESCRIPTION
**Pull request by bacongobbler**
This fixes an issue where capabilities that the actor did not add to its claim are being added, causing a runtime error. We should only be binding capabilities that the actor requested.

closes #239

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

### CODE REVIEW

File: crates/wascc-provider/src/lib.rs [hunk 1]
```diff
@@ -650,28 +627,44 @@ impl kubelet::log::HandleFactory<tokio::fs::File> for LogHandleFactory {
...
+            env: logenv,
+        });
+    }
+
+    if actor_caps.contains(&HTTP_CAPABILITY.to_owned()) {
+        env.insert("PORT".to_string(), port_assigned.to_string());
+        capabilities.push(Capability {
+            name: HTTP_CAPABILITY,
+            binding: None,
+            env,
```
Comment (thomastaylor312) [inline crates/wascc-provider/src/lib.rs:663]: So this is a little janky (not due to your code). For the actor to get the env variables that are assigned in Kubernetes, I believe those variables need to be assigned to the capability. So as it stands right now, the containers in the pod will only get the environment variables if they have the http capability.

Comment (bacongobbler) [reply crates/wascc-provider/src/lib.rs:663]: Can yo clarify what needs to be addressed here to fix this? Is there some test case I can use to verify this issue?

Comment (thomastaylor312) [reply crates/wascc-provider/src/lib.rs:663]: The `env` hash map comes from the environment variables defined in the pod spec. We expect these variables to be set within the "container." These variables get passed to the actor when binding the provider. So if a provider doesn't support the http capability, it won't have the environment variables from the pod definition. You can test this by running an actor that prints out an environment variable specified in the pod spec, but doesn't have the http capability. You should see that the env variable is empty when it prints

Comment (bacongobbler) [reply crates/wascc-provider/src/lib.rs:663]: Okay. I think I see where you're coming from now. I'll see what I can do here, and see if I can write an integration test using @itowlson's wasmerciser.

Comment (bacongobbler) [reply crates/wascc-provider/src/lib.rs:663]: Reading the source code more closely, this seems intentional from a design standpoint. Each call to `bind_actor` contains a `HashMap<String, String>` that defines the configuration used between the actor and the capability. As far as I'm aware of, there is no concept of an environment variable in waSCC.

In other words, right now we're binding every environment variable in the pod manifest as configuration for the http server capability, and right now the [only value the http server capability cares about is PORT](https://github.com/wascc/http-server-provider/blob/7d538bb732f56e4e5dc1303c89cbce5515e72f7c/src/lib.rs#L83).

In other words, any environment variables you add to the pod manifest are currently ignored or unused. I'll open a new ticket to investigate how we should handle these going forward. Does that sound good to you?

Comment (thomastaylor312) [reply crates/wascc-provider/src/lib.rs:663]: That sounds fine to me. We can tackle a solution in 0.5
